designs/v1beta1-full-changelist.md

```diff
@@ -0,0 +1,200 @@
+
+If `consolidationPolicy` is not set, Karpenter will implicitly default to `WhenUnderutilized`.
+
+#### `provisioner.Spec.ttlSecondsAfterEmpty` → `nodePool.spec.deprovisioning.consolidationTTL`
+
+While the `consolidationPolicy` offers one mechanism for users to control the aggressiveness of deprovisioning, users that enable a `consolidationPolicy` of `WhenEmpty` or `WhenUnderutilized` may still want to dictate the speed at which nodes are deemed underutilized. This is particularly true on clusters that are large in size and have a large amount of churn. To support this, Karpenter will surface a `consolidationTTL` which will allow you to define a per-node TTL to define the time that Karpenter can begin deprovisioning the node after first seeing that the node is eligible for consolidation.
+
+#### Remove `nodePool.spec.provider`
+
+We’ve recommended that customers leverage spec.providerRef in favor of spec.provider since Q2 2022. Documentation for this feature has been removed since Q3 2022. We will take the opportunity to remove the feature entirely to minimize code bugs/complexity and user confusion.
+
+### `NodeClass` Changes
+
+#### Update `nodeClass.spec.amiSelector`
+
+The current `amiSelector` has two primary limitations that restrict user’s ability to specify the AMIs that they want Karpenter to use:
+
+1. Users can only specify “ANDed” together requirements, meaning that if a user has an orthogonal set of tags that they want to match their images to, they have to specify them by `aws::ids` directly, since there is no way with the current tag-selection logic to specify those values
+2. Users want more flexibility to do things like specify a name/owner combination for images. Users have generally been asking Karpenter to more closely adhere to the EC2 APIs in our amiSelector design so that users can use more built-in filtering for AMIs, instead of having to use custom tagging to achieve the same outcome
+    1. To support *some* of these use-cases, Karpenter has begun effectively creating “system-tags” i.e. (`aws::ids`, `aws::owners`, `aws::name`). These are special-cased version of the standard user custom-tags that allow users to achieve the scenarios described in #2; however, they are not easily discoverable or understood and if we are beginning to support special-cases like this, it makes sense that we should begin to structure these fields.
+
+```
+amiSelector:
+- name: foo
+  id: abc-123
+  owner: amazon
+  ssm: "/al2/alias"
+  tags: 
+    key: value
+# Selectors are ORed
```

Comment:
Not in Kubernetes they're not! I'd avoid calling this a selector.

Reply:
Good callout. Is this the reason that they are called `NodeSelectorTerms` in [`nodeAffinity`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#nodeselector-v1-core:~:text=NodeSelectorTerm%20array)? Maybe we should model this the same way with `amiSelectorTerms`?

Reply:
Even `nodeSelectorTerms` is an AND expression. I would use a word other than selector, especially as the backing API doesn't look (or work) like a Kubernetes one.

If we call it a selector, people who know a different mental model - ie, actual selectors - may assume it works just like those ones.
There's a related concern. Imagine that Kubernetes one day adds a logical-OR mechanism. Once we stabilise this API it's difficult to revise it, so the Karpenter “selectors“ might be stuck working similarly to actual selectors, but frustratingly not the same.

Reply:
As a thought experiment, try expressing these requirements first using EC2's `Filter` API and then using a Kubernetes-style selector:

- AMI must **not** be public
- AMI must be for AMD64 CPU architecture
- AMI must have a tag set on it, `example.com/kubelet-minor-version`
- Volume type must be `gp2` or `gp3`

I think that's a reasonable filter and it's one you can specify using the underlying API. So, we should consider allowing it. But we can't make that API look enough like a Kubernetes selector (even the [set-based requirement](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#set-based-requirement) kind), because there are different ergonomics here. The problem actually isn't the OR expression, now I think about it; it's what we set the keys to. Actual node selectors rely on automatic labelling, but I'm worried about whether it's feasible to map AWS tags to look like label keys.

---

Calling it a filter (and, to be clear, automating the migration for people using the older API) avoids people having to learn the subtle traps here.

More importantly, I think: a future change to the EC2 `DescribeImages` operation might make us want to expose more ways to filter images. The more that we try to adapt that API to look like Kubernetes, when it isn't really, the bigger the risk of future trouble - and problems that AWS' internal API change review process is unlikely to account for.

Reply:
> Even `nodeSelectorTerms` is an AND expression

The [K8s documentation](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#:~:text=NodeSelectorRequirement%20v1%20core) mentions that these terms are ORed

<img width="680" alt="Screenshot 2023-07-30 at 10 35 36 PM" src="https://github.com/aws/karpenter/assets/26334334/7af02f4a-39cf-4b3c-bcc3-d3a2e28775fe">

Reply:
> I'm worried about whether it's feasible to map AWS tags to look like label keys

I think, generally, the user has complete control over these tags (if they are using tag-based selection) so ideally we shouldn't hit conflicts with label selection mechanisms in Kubernetes. I'm still leaning towards calling it a `selector` with terms since it's more K8s-native and the purpose of these API fields is a "selection"-based mechanism

Reply:
I do tech writing (and cloud consultancy) and I try to avoid APIs where you end up having to write things like “**NOTE** This looks like a selector in Kubernetes but isn't the same, here are a list of differences that might catch you out”.

Given that experience sometimes I look upstream to see where hard-to-document things are popping up. That's all the context I can offer, anyway. I think this will be hard to document and I worry that there will be a steady stream of slightly confused people adopting this API.

Reply:
> This looks like a selector in Kubernetes but isn't the same, here are a list of differences that might catch you out

I'm slightly confused by this statement. I'm missing what exactly is different about this mechanism other than the service that it is interacting with. In all other ways, it should behave the same as selectors that interact directly with the K8s API server.